src/storage/txn/process.rs

```diff
@@ -444,25 +524,84 @@ fn process_write_impl<S: Snapshot>(
             let mut txn = MvccTxn::new(snapshot, start_ts, !ctx.get_not_fill_cache())?;
             let mut locks = vec![];
             let rows = mutations.len();
-            for m in mutations {
-                match txn.prewrite(m, &primary, &options) {
+            for (i, m) in mutations.into_iter().enumerate() {
+                // If `options.is_pessimistic_lock` is empty, the transaction is optimistic
+                // or else pessimistic.
+                if options.is_pessimistic_lock.is_empty() {
```

Comment:
Could we check `is_pessimistic_lock` outside the loop? Maybe before L527

Reply:
Yes